book/chrome.md

```diff
@@ -8,531 +8,797 @@ next: forms
+other types of clickable things later---like text boxes and buttons in
+the [next chapter](forms.md). Finally, now that we've found the link
+element itself, we need to extract the URL and direct the browser to
+it. This URL might be a relative URL:
+
+``` {.python indent=12}
+# ...
+elif elt.tag == "a" and "href" in elt.attributes:
+    url = relative_url(elt.attributes["href"], self.url)
+    return self.load(url)
+```
+
+Since this needs to know the browser's current URL, we need to store
+it in `load`:
+
+``` {.python replace=Browser/Tab}
+class Browser:
+    def load(self, url):
+        self.url = url
         # ...
-        self.children.pop()
 ```
 
-Now that words and lines lay themselves out, a lot of stuff disappears
-from`InlineLayout`. The `style`, `size`, `weight` fields are no longer
-needed. Neither is `line` or `cx`. And the display list variable is no
-longer needed: instead of an `InlineLayout` drawing all the words in a
-paragraph, each word can now be drawn by its corresponding a
-`TextLayout` object:
+Try it! You should now be able to click on links and navigate to new
+web pages.
+
+Multiple pages
+==============
+
+If you're anything like me, the next thing you tried after clicking on
+links is middle-clicking them to open in a new tab. Every browser now
+has tabbed browsing, and honestly it's a little embarrasing that our
+little toy browser doesn't.
+
+Fundamentally tabbed browsing means we need to distinguish between the
+browser itself and individual tabs that browse some specific web page.
+Right now the `Browser` class stores both information about the
+browser (like the canvas it draws to) and information about a single
+tab (like the layout tree and display list). We need to tease the two
+apart.
+
+Here's the plan: the `Browser` class will store the window and canvas,
+plus the list of tabs. Everything else goes into a new `Page` class.
+Since the `Browser` stores the window and canvas, it handles all of
+the events, sometimes forwarding it to the active tab.
+
+So the `Tab` constructor looks like this:
 
 ``` {.python}
-class TextLayout:
-    def paint(self, to):
-        color = self.node.style["color"]
-        to.append(DrawText(self.x, self.y, self.word, self.font, color))
+class Tab:
+    def __init__(self):
+        with open("browser6.css") as f:
+            self.default_style_sheet = CSSParser(f.read()).parse()
 ```
 
-The `InlineLayout` and `LineLayout` versions of the `draw` method
-now don't need to do anything but recurse on their children.
+The `load`, `scrolldown`, `click`, and `draw` methods all move to
+`Tab`, since that's now where all web-page-specific data lives. But
+there's a change here too.
 
-Navigating between pages
-========================
+Since the `Browser` controls the canvas and handles events, it decides
+when rendering happens and which tab does the drawing. After all, you
+only want one tab drawing its contents at a time! So let's remove the
```

Comment:
```suggestion
only want one tab drawing its contents at a time![^windowed] So let's remove the
```

Reply:
Plus:

[^windowed] Unless the browser implements multiple windows!

And add an exercise for this.